numpy/core/src/multiarray/einsum.c.src

```diff
@@ -2769,8 +2769,13 @@ PyArray_EinsteinSum(char *subscripts, npy_intp nop,
 
     /* Initialize the output to all zeros and reset the iterator */
     ret = NpyIter_GetOperandArray(iter)[nop];
+    if (PyArray_AssignZero(ret, NULL) < 0) {
+        goto fail;
+    }
+    if (out != NULL) {
```

Comment:
This looks fishy, why do anything to `ret` if it is simply set to `out`?

Reply:
If the output operand has writeback semantics, `NpyIter_GetOperandArray(iter)[nop]` is the temporary ndarray created with `out` as it's base ndarray. Then we zero the temporary data rather than changing `out` itself. See issue #10956 for the python-level equivalent discussion about what is meant by `iter.operand[i]` - the temporary ndarray with writeback semantics or the underlying user-created input-ouput ndarray.

Lines 2775-2777 ensure that `ret` and `out` point to the same ndarray, as required when using an `out` kwarg.